app/components/billing/payment.js

```diff
@@ -53,6 +60,38 @@ export default Component.extend({
     }
   }).drop(),
 
+  // amount_off and price are in cents
+  discountedPrice: computed('coupon.{amountOff,percentOff}', 'selectedPlan.price', function () {
+    const price = Math.floor(this.selectedPlan.price / 100);
+    if (this.coupon && this.coupon.amountOff) {
+      const { amountOff } = this.coupon;
+      const discountedPrice = price - Math.floor(amountOff / 100);
+      return `$${discountedPrice}`;
+    } else if (this.coupon && this.coupon.percentageOff) {
+      const { percentageOff } = this.coupon;
+      const discountedPrice = price - (price * percentageOff) / 100;
+      return `$${discountedPrice.toFixed(2)}`;
+    } {
+      return `$${price}`;
+    }
+  }),
```

Comment:
No, service is not the right place for such things. Services are intended to share global state within application. These fields are record-specific, meaning they are different for different records. Let's say that we have two subscriptions in the store. Which values would this service contain in that case? Switching subscription for the service to calculate values is a bad idea, because service maintains its state globally (a single instance for the entire application), which means that switching subscription in the service in one place would affect all other places that use this service. Does it make sense?

That said, these properties are record-specific, and thus, they should belong to model. And I'm not entirely sure, but looking at the [subscription model](https://github.com/travis-ci/travis-web/blob/master/app/models/subscription.js) it seems like they belong there, not to `plan`. These properties depend on `coupon` and `price`. Coupon [belongs to subscription](https://github.com/travis-ci/travis-web/blob/master/app/models/subscription.js#L23) and price is available via [`plan`](https://github.com/travis-ci/travis-web/blob/master/app/models/subscription.js#L31).[`price`](https://github.com/travis-ci/travis-web/blob/master/app/models/plan.js#L6). And also, thinking about the meaning - when we apply a coupon, we don't decrease _plan's_ price, but we make a discount for this specific subscription, right? It makes me think, that `totalPrice` should belong to `subscription`, not to `plan`.

> subscription hasn't been created at the time of applying a coupon so the subscription won't have a price
Looking through the code, it seems like the `newSubscription` is supplied as an attribute to this component. And once the plan is selected, it should have `plan` relation set and `price` would be accessible through it.

I might be missing certain points and data relations - please let me know if that's the case. These are basically my assumptions from the code and common sense. 

What do you think?